src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, Arm Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "asm/macroAssembler.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#include "classfile/symbolTable.hpp"
+#include "include/jvm.h"
+#include "jni.h"
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "memory/allocation.inline.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/arrayOop.inline.hpp"
+#include "prims/universalUpcallHandler.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/jniHandles.inline.hpp"
+
+static constexpr CodeBuffer::csize_t upcall_stub_size = 1024;
+
+extern struct JavaVM_ main_vm;
+
+static struct {
+  bool inited;
+  struct {
+    Klass* klass;
+    Symbol* name;
+    Symbol* sig;
+  } upcall_method;  // jdk.internal.foreign.abi.UniversalUpcallHandler::invoke
+} upcall_info;
+
+// FIXME: This should be initialized explicitly instead of lazily/racily
+static void upcall_init() {
+#if 0
+  fprintf(stderr, "upcall_init()\n");
+#endif
+
+  TRAPS = Thread::current();
+  ResourceMark rm;
+
+  const char* cname = "jdk/internal/foreign/abi/ProgrammableUpcallHandler";
+  const char* mname = "invoke";
+  const char* mdesc = "(Ljdk/internal/foreign/abi/ProgrammableUpcallHandler;J)V";
+  Symbol* cname_sym = SymbolTable::new_symbol(cname, (int)strlen(cname));
+  Symbol* mname_sym = SymbolTable::new_symbol(mname, (int)strlen(mname));
+  Symbol* mdesc_sym = SymbolTable::new_symbol(mdesc, (int)strlen(mdesc));
```

Comment:
You don't need the strlen() argument.

Reply:
Ok, I see it has an overload that does that in the header file (and I only looked at the .cpp file. :( )